src/cmd/services/m3comparator/main/series_iterator_builder.go

```diff
@@ -0,0 +1,161 @@
+// Copyright (c) 2019 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package main
+
+import (
+	"io"
+	"strings"
+	"time"
+
+	"github.com/m3db/m3/src/dbnode/encoding"
+	"github.com/m3db/m3/src/dbnode/encoding/m3tsz"
+	"github.com/m3db/m3/src/dbnode/namespace"
+	"github.com/m3db/m3/src/dbnode/x/xio"
+	"github.com/m3db/m3/src/x/ident"
+	xtime "github.com/m3db/m3/src/x/time"
+)
+
+const sep rune = '!'
+const tagSep rune = '.'
+
+type iteratorOptions struct {
+	blockSize     time.Duration
+	start         time.Time
+	encoderPool   encoding.EncoderPool
+	iteratorPools encoding.IteratorPools
+}
+
+var iterAlloc = func(r io.Reader, _ namespace.SchemaDescr) encoding.ReaderIterator {
+	return m3tsz.NewReaderIterator(r, m3tsz.DefaultIntOptimizationEnabled, encoding.NewOptions())
+}
+
+func buildBlockReader(
+	block seriesBlock,
+	start time.Time,
+	opts iteratorOptions,
+) ([]xio.BlockReader, error) {
+	encoder := opts.encoderPool.Get()
+	encoder.Reset(start, len(block), nil)
+	for _, dp := range block {
+		err := encoder.Encode(dp, xtime.Second, nil)
+		if err != nil {
+			encoder.Close()
+			return nil, err
+		}
+	}
+
+	segment := encoder.Discard()
+	encoder.Close()
+	return []xio.BlockReader{
+		xio.BlockReader{
+			SegmentReader: xio.NewSegmentReader(segment),
+			Start:         start,
+			BlockSize:     opts.blockSize,
+		},
+	}, nil
+}
+
+func buildTagIteratorAndID(tagMap tagMap) (ident.TagIterator, ident.ID) {
+	var (
+		tags = ident.Tags{}
+		sb   strings.Builder
+	)
+
+	for name, value := range tagMap {
```

Comment:
Since for-looping over a map, won't the order not be deterministic?

Would this at some point cause problems?

Seems like just building with `models.Tags` and then outputting (which will sort it yeah?) would cause less problems if this ever becomes important later.

Reply:
Yeah fair enough, will refactor